arangod/Aql/HashedCollectExecutor.cpp

```diff
@@ -242,55 +243,41 @@ std::pair<ExecutionState, NoStats> HashedCollectExecutor::produceRow(OutputAqlIt
   return {state, NoStats{}};
 }
 
-// if no group exists for the current row yet, this builds a new group.
-std::pair<std::unique_ptr<HashedCollectExecutor::AggregateValuesType>, std::vector<AqlValue>>
-HashedCollectExecutor::buildNewGroup(InputAqlItemRow& input, size_t n) {
-  GroupKeyType group;
-  group.reserve(n);
-
-  // copy the group values before they get invalidated
-  for (size_t i = 0; i < n; ++i) {
-    group.emplace_back(input.stealValue(_infos.getGroupRegisters()[i].second));
-  }
-
-  auto aggregateValues = std::make_unique<AggregateValuesType>();
-  aggregateValues->reserve(_aggregatorFactories.size());
-
-  for (auto const& it : _aggregatorFactories) {
-    aggregateValues->emplace_back((*it)(_infos.getTransaction()));
-  }
-  return std::make_pair(std::move(aggregateValues), group);
-}
-
 // finds the group matching the current row, or emplaces it. in either case,
 // it returns an iterator to the group matching the current row in
 // _allGroups. additionally, .second is true iff a new group was emplaced.
 decltype(HashedCollectExecutor::_allGroups)::iterator HashedCollectExecutor::findOrEmplaceGroup(
     InputAqlItemRow& input) {
-  GroupKeyType groupValues;  // TODO store groupValues locally
-  size_t const n = _infos.getGroupRegisters().size();
-  groupValues.reserve(n);
+  _nextGroupValues.clear();
 
   // for hashing simply re-use the aggregate registers, without cloning
   // their contents
-  for (size_t i = 0; i < n; ++i) {
-    groupValues.emplace_back(input.getValue(_infos.getGroupRegisters()[i].second));
+  for (auto const& reg : _infos.getGroupRegisters()) {
+    _nextGroupValues.emplace_back(input.getValue(reg.second));
   }
 
-  auto it = _allGroups.find(groupValues);
+  auto it = _allGroups.find(_nextGroupValues);
 
   if (it != _allGroups.end()) {
     // group already exists
     return it;
   }
 
   // must create new group
-  GroupValueType aggregateValues;
   GroupKeyType group;
-  std::tie(aggregateValues, group) = buildNewGroup(input, n);
+  group.reserve(_infos.getGroupRegisters().size());
+  group.swap(_nextGroupValues);
+
+  // this builds a new group with aggregate functions beeing prepared.
+  auto aggregateValues = std::make_unique<AggregateValuesType>();
+  aggregateValues->reserve(_aggregatorFactories.size());
+  auto trx = _infos.getTransaction();
+  for (auto const& it : _aggregatorFactories) {
+    aggregateValues->emplace_back((*it)(trx));
+  }
 
   // note: aggregateValues may be a nullptr!
-  auto emplaceResult = _allGroups.emplace(group, std::move(aggregateValues));
+  auto emplaceResult = _allGroups.emplace(std::move(group), std::move(aggregateValues));
```

Comment:
You could just put
```suggestion
  auto emplaceResult = _allGroups.emplace(std::move(_nextGroupValues), std::move(aggregateValues));
```
here and omit the `group` variable. After that, `_nextGroupValues` will be empty.